### PR TITLE
#913 uncompilable code generated in 1.1.0.CR1 for collection properties

### PR DESCRIPTION
**Pull request by sjaakd**
The solution is twofold ( 1 & 2) but with some consequences ( 3 & 4):
1. It introduces the new ideas I have for templates. It uses the `<#nested>` in Freemarker to wrap statements in a null check or try-catch clause. The idea is to gradually phase out the other `xyzWrapper` and only keep an lefthandside (target) and righthandside (source), making stuff a lot simpeler.
2. Delegate the `NullValueMappingStrategy` to the forged Map/Iterable mapping methods and let them handle default situations. This requires that the user is warned for creating plausible null pointer situations. 
3. Choose (in case the user did not define a `NullValueMappingStrategy`  explicitly for forged methods the `NullValueMappingStrategy.RETURN_DEFAULT` iso  `NullValueMappingStrategy.RETURN_NULL`. This seems more in line with the idea that a used mapping method itself is responsible for handling null input.
4. There's a drawback, a testcase (#865) starts to fail that expects null to be returned when both source and target collection, which seems reasonable. Possibly bring back some null checks on source here.

Next step (after CR2): 

a. rebasing this on #903 it should be simple to get rid of the `NewCollectionOrMapWrapper` and `EnumSetCopyWrapper`. 

b. As further optimization: a local variable is not required when there's no NPE check.

c. Plain property mappings in a similar way?

There seems to be an issue with delegating the `NullValueMappingStrategy` from `@Mapper` to `@BeanMapping`

### CODE REVIEW

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_289/Issue289Mapper.java [hunk 1]
```diff
@@ -31,11 +33,9 @@
 
     void sourceToTargetWithoutSetter(Source source, @MappingTarget TargetWithoutSetter target);
 
+    @BeanMapping( nullValueMappingStrategy = NullValueMappingStrategy.RETURN_NULL )
```
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_289/Issue289Mapper.java]: should not be required anymore

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 2]
```diff
@@ -681,7 +675,8 @@ private Assignment forgeMapMapping(Type sourceType, Type targetType, SourceRHS s
         private String getName(Type sourceType, Type targetType) {
             String fromName = getName( sourceType );
             String toName = getName( targetType );
-            return Strings.decapitalize( fromName + "To" + toName );
+            String with = nullValueMappingStrategy != null ? "With" + nullValueMappingStrategy.name() : "";
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: The generated name follows the enum constant. So capitals and underscore.. Should I decapitilize?

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 3]
```diff
@@ -394,6 +399,19 @@ private void useLocalVarWhenNested(Assignment rightHandSide) {
             }
         }
 
+        private String getSafeName() {
+            String name;
+            if ( sourceReference.getPropertyEntries().isEmpty() ) {
+                name = sourceReference.getParameter().getName();
+            }
+            else {
+                name = sourceReference.getPropertyEntries().get( 0 ).getName();
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: use first()

**Comment by sjaakd** [comment]
I'll process my own comments together with the others (when there).

**Comment by gunnarmorling** [comment]
> rebasing this on #903 it should be simple to get rid of the NewCollectionOrMapWrapper and EnumSetCopyWrapper. 

Do you mean it the other way around I reckon? Rebase #903 on top of this?

**Comment by gunnarmorling** [comment]
@sjaakd do you think you could share an example of the generated code (and how it differs from the previous one?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_289/Issue289Test.java [hunk 4]
```diff
@@ -79,4 +82,46 @@ public void shouldLeaveNullTargetSetWhenSourceIsNullForCreateMethod() {
...
     }
+
+    @Test
+    @ExpectedCompilationOutcome(
+        value = CompilationResult.SUCCEEDED,
+        diagnostics = {
+            @Diagnostic(type = Issue289MapperTargetOnlyGetterWithWarning.class,
+                kind = javax.tools.Diagnostic.Kind.WARNING,
+                line = 33,
+                messageRegExp = "NullPointer risk when only a read accessor is available and applying "
```
Comment (gunnarmorling) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_289/Issue289Test.java]: This one makes me curious. Maybe we can talk about it in the call tonight?

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 5]
```diff
@@ -681,7 +676,9 @@ private Assignment forgeMapMapping(Type sourceType, Type targetType, SourceRHS s
         private String getName(Type sourceType, Type targetType) {
             String fromName = getName( sourceType );
             String toName = getName( targetType );
-            return Strings.decapitalize( fromName + "To" + toName );
+            String with = nullValueMappingStrategy != null ?
+                "WithNvms" + camelizeEnumConstant( nullValueMappingStrategy ) : "";
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: "camelize"... that's flowerly... 🐫 👍

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 6]
```diff
@@ -692,6 +689,43 @@ private String getName(Type type) {
...
+
+        private NullValueMappingStrategyPrism determineForgedNullValueMappingStrategy() {
+
+            NullValueMappingStrategyPrism nvms;
+
+            warnWhenTargetWriteAccessorIsGetterAndNullValueMappingIsExplicitReturnNull();
+            warnWhenUpdateMethodAndNullValueMappingIsExplicitReturnNull();
+
+            if ( nullValueMappingStrategy == null
+                && ( method.isUpdateMethod() || TargetWriteAccessorType.GETTER.equals( targetWriteAccessorType ) ) ) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: use `==` for enum equality check, please... 😃

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 7]
```diff
@@ -692,6 +689,43 @@ private String getName(Type type) {
...
+            else {
+                nvms = nullValueMappingStrategy;
+            }
+            return nvms;
+        }
+
+        private void warnWhenTargetWriteAccessorIsGetterAndNullValueMappingIsExplicitReturnNull() {
+
+            if ( TargetWriteAccessorType.GETTER.equals( targetWriteAccessorType )
+                && NullValueMappingStrategyPrism.RETURN_NULL.equals( nullValueMappingStrategy ) ) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: use `==` for enum equality check, please... 😃

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 8]
```diff
@@ -692,6 +689,43 @@ private String getName(Type type) {
...
+                ctx.getMessager().printMessage( method.getExecutable(),
+                    Message.PROPERTYMAPPING_POSSIBLE_NULL_VALUE_WHEN_NO_WRITE_ACCESSOR_TYPE,
+                    targetPropertyName );
+            }
+        }
+
+        private void warnWhenUpdateMethodAndNullValueMappingIsExplicitReturnNull() {
+
+            if ( method.isUpdateMethod()
+                && NullValueMappingStrategyPrism.RETURN_NULL.equals( nullValueMappingStrategy ) ) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: use `==` for enum equality check, please... 😃

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/GetterWrapperForCollectionsAndMaps.java [hunk 9]
```diff
@@ -74,20 +65,18 @@ public GetterWrapperForCollectionsAndMaps(Assignment decoratedAssignment, List<T
...
-    }
-
-    public String getLocalVarName() {
-        return localVarName;
+    @Override
+    public boolean isIncludeNullCheck() {
+        // For collections and maps include a null check, when the assignment type is DIRECT.
+        // for mapping methods (builtin / custom), the mapping method is responsible for the
+        // null check. Typeconversions do not apply to collections and maps.
+        return AssignmentType.DIRECT.equals( getType() );
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/GetterWrapperForCollectionsAndMaps.java]: use `==` for enum equality check, please... 😃

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java [hunk 10]
```diff
@@ -65,17 +63,35 @@ public Assignment getNewCollectionOrMapAssignment() {
...
     }
 
-    public String getLocalVarName() {
-        return localVarName;
+    @Override
+    public boolean isIncludeNullCheck() {
+        // For collections and maps include a null check, when the assignment type is DIRECT.
+        // for mapping methods (builtin / custom), the mapping method is responsible for the
+        // null check. Typeconversions do not apply to collections and maps.
+        return AssignmentType.DIRECT.equals( getType() );
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java]: use `==` for enum equality check, please... 😃

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 11]
```diff
@@ -53,6 +54,8 @@
     PROPERTYMAPPING_INVALID_PROPERTY_NAME( "No property named \"%s\" exists in source parameter(s)." ),
     PROPERTYMAPPING_NO_PRESENCE_CHECKER_FOR_SOURCE_TYPE( "Using custom source value presence checking strategy, but no presence checker found for %s in source type." ),
     PROPERTYMAPPING_NO_READ_ACCESSOR_FOR_TARGET_TYPE( "No read accessor found for property \"%s\" in target type." ),
+    PROPERTYMAPPING_POSSIBLE_NULL_VALUE_WHEN_NO_WRITE_ACCESSOR_TYPE( "NullPointer risk when only a read accessor is available and applying NullValueMappingStrategy.RETURN_NULL for property \"%s\" in target type.", Diagnostic.Kind.WARNING ),
+    PROPERTYMAPPING_POSSIBLE_NULL_VALUE_UPDATE_METHOD_ACCESSOR_TYPE( "NullPointer risk when using update method and applying NullValueMappingStrategy.RETURN_NULL for property \"%s\" in target type.", Diagnostic.Kind.WARNING ),
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Could you give us an example of what exactly that means? It could perhaps be something for our to-be-created FAQ page...

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_289/Issue289Mapper.java [hunk 12]
```diff
@@ -35,7 +35,4 @@
 
     TargetElement sourceElementToTargetElement(SourceElement source);
 
-
-
-
```
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_289/Issue289Mapper.java]: 👍

File: processor/src/main/resources/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.ftl [hunk 13]
```diff
@@ -18,41 +18,41 @@
...
+<#import "../macro/SourceUtil.ftl" as source>
+<#import "../macro/TargetUtil.ftl" as target>
+<@source.handleExceptions>
 <#if ( ext.existingInstanceMapping ) >
-    <@_assignment targetWriteAccessorName = localVarName/>
     if ( ${ext.targetBeanName}.${targetGetterName}() != null ) {
         ${ext.targetBeanName}.${targetGetterName}().clear();
-        if ( ${localVarName} != null ) {
-            ${ext.targetBeanName}.${ext.targetReadAccessorName}().<#if ext.targetType.collectionType>addAll<#else>putAll</#if>( ${localVarName} );
+        <@source.handleNullCheck>
```
Comment (agudian) [inline processor/src/main/resources/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.ftl]: That's pretty cool! 👍

**Comment by sjaakd** [comment]
> Do you mean it the other way around I reckon? Rebase #903 on top of this?

Yes.. First this one.. Later the `AdderWrapper` improvements. Then adapt this accordingly. Since this one solves an issue (reproducer included), we can include this in the CR. The remainder is for later...

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 11 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: There are two new test cases in the issue #289 reproducer. I will move them (in due time) to collection testing. Do they answer your question? I can of course externalize them..

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java (hunk 10 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java]: Ok..

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java (hunk 5 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: :smile: .. Sounds like caramelize..

**Comment by sjaakd** [comment]
> I don't yet get how that'll help us with our customizability. But it sure makes the code a bit more readable.

I have some ideas on that area. But first make stuff understandable for ourselves again. If that works, I'm confident I can make the next move. Step by step..

**Comment by sjaakd** [comment]
> @sjaakd do you think you could share an example of the generated code (and how it differs from the previous one?

The main problem was this one:

old:

```java
public class DomainDtoMapperImpl implements DomainDtoMapper {

    @Override
    public Domain update(Dto source, Domain target) {
        if ( source == null ) {
            return null;
        }

        if ( source.getStrings() != null ) { // strange double check
            Set<String> set = source.getStrings(); // strange assignment
            if ( target.getStrings() != null ) {
                target.getStrings().clear();
                if ( set != null ) {
                    target.getStrings().addAll( set );
                }
            }
            else {
                target.setStrings( new HashSet<String>( source.getStrings() ) );
            }
        }

        return target;
    }
}
```
 new:

```java
public class DomainDtoMapperImpl implements DomainDtoMapper {

    @Override
    public Domain update(Dto source, Domain target) {
        if ( source == null ) {
            return null;
        }

        if ( target.getStrings() != null ) {
            target.getStrings().clear();
            List<String> strings = source.getStrings();
            if ( strings != null ) {
                target.getStrings().addAll( strings );
            }
        }
        else {
            List<String> strings = source.getStrings();
            if ( strings != null ) {
                target.setStrings( new HashSet<String>( strings ) );
            }
        }

        return target;
    }
}
```

Furthermore, I decided that we could delegate the null value mapping strategy to the forged iterable / map mapping method. We do this for other annotation attributes as well (I remember a PR of some weeks ago).

By default (no annotation provided): we would select `RETURN_DEFAULT` for calls from update / getter methods and "RETURN_NULL" for calls from a setter method. 

On the call site, (in the generated code) this saves an NPE check and is in line with the idea that mapping methods should provide a decent null handling. Hence, if the warning if the user overrides this.

So take for instance  the reproducer of the old #289 issue.

old:

```java
public class Issue289MapperImpl implements Issue289Mapper {

    @Override
    public TargetWithoutSetter sourceToTargetWithoutSetter(Source source) {
        if ( source == null ) {
            return null;
        }

        TargetWithoutSetter targetWithoutSetter = new TargetWithoutSetter();

        if ( targetWithoutSetter.getCollection() != null ) {
            Collection<TargetElement> targetCollection = sourceElementCollectionToTargetElementCollection( source.getCollection() );
            if ( targetCollection != null ) {
                targetWithoutSetter.getCollection().addAll( targetCollection );
            }
        }

        return targetWithoutSetter;
    }

    @Override
    public void sourceToTargetWithoutSetter(Source source, TargetWithoutSetter target) {
        if ( source == null ) {
            return;
        }

        if ( target.getCollection() != null ) {
            target.getCollection().clear();
            Collection<TargetElement> targetCollection = sourceElementCollectionToTargetElementCollection( source.getCollection() );
            if ( targetCollection != null ) { // check
                target.getCollection().addAll( targetCollection );
            }
        }
    }

    @Override
    public TargetWithSetter sourceToTargetWithSetter(Source source) {
        if ( source == null ) {
            return null;
        }

        TargetWithSetter targetWithSetter = new TargetWithSetter();

        targetWithSetter.setCollection( sourceElementCollectionToTargetElementCollection( source.getCollection() ) ); 

        return targetWithSetter;
    }

    @Override
    public TargetElement sourceElementToTargetElement(SourceElement source) {
        if ( source == null ) {
            return null;
        }

        TargetElement targetElement___ = new TargetElement();

        targetElement___.setValue( source.getValue() );

        return targetElement___;
    }

    protected Collection<TargetElement> sourceElementCollectionToTargetElementCollection(Collection<SourceElement> collection) {
        if ( collection == null ) {
            return null; // null is returned always
        }

        Collection<TargetElement> collection_ = new ArrayList<TargetElement>();
        for ( SourceElement sourceElement : collection ) {
            collection_.add( sourceElementToTargetElement( sourceElement ) );
        }

        return collection_;
    }
}
```

new:

``` java
public class Issue289MapperImpl implements Issue289Mapper {

    @Override
    public TargetWithoutSetter sourceToTargetWithoutSetter(Source source) {
        if ( source == null ) {
            return null;
        }

        TargetWithoutSetter targetWithoutSetter = new TargetWithoutSetter();

        if ( targetWithoutSetter.getCollection() != null ) {
            Collection<SourceElement> collection = source.getCollection();
            targetWithoutSetter.getCollection().addAll( sourceElementCollectionToTargetElementCollection( collection ) );
        }

        return targetWithoutSetter;
    }

    @Override
    public void sourceToTargetWithoutSetter(Source source, TargetWithoutSetter target) {
        if ( source == null ) {
            return;
        }

        if ( target.getCollection() != null ) {
            Collection<SourceElement> collection = source.getCollection();
            target.getCollection().clear();
            target.getCollection().addAll( sourceElementCollectionToTargetElementCollection( collection ) ); // MapStruct will generate different forged methods depending on whether this is an update method / target dto without setter. This forged method will return an empty array on null input.
        }
    }

    @Override
    public TargetWithSetter sourceToTargetWithSetter(Source source) {
        if ( source == null ) {
            return null;
        }

        TargetWithSetter targetWithSetter = new TargetWithSetter();

        Collection<SourceElement> collection = source.getCollection();
        targetWithSetter.setCollection( sourceElementCollectionToTargetElementCollection_( collection ) ); // MapStruct will generate different forged methods depending on whether this is an update method / target dto without setter. This forged method will return null on null input.

        return targetWithSetter;
    }

    @Override
    public TargetElement sourceElementToTargetElement(SourceElement source) {
        if ( source == null ) {
            return null;
        }

        TargetElement targetElement___ = new TargetElement();

        targetElement___.setValue( source.getValue() );

        return targetElement___;
    }

    protected Collection<TargetElement> sourceElementCollectionToTargetElementCollection(Collection<SourceElement> collection) {
        if ( collection == null ) {
            return new ArrayList<TargetElement>();
        }

        Collection<TargetElement> collection_ = new ArrayList<TargetElement>();
        for ( SourceElement sourceElement : collection ) {
            collection_.add( sourceElementToTargetElement( sourceElement ) );
        }

        return collection_;
    }

    protected Collection<TargetElement> sourceElementCollectionToTargetElementCollection_(Collection<SourceElement> collection) {
        if ( collection == null ) {
            return null;
        }

        Collection<TargetElement> collection_ = new ArrayList<TargetElement>();
        for ( SourceElement sourceElement : collection ) {
            collection_.add( sourceElementToTargetElement( sourceElement ) );
        }

        return collection_;
    }
}
```

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java (hunk 10 above)
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java]: Seems Andreas got stuck in a time loop ;)

Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java]: a twist.. in the fabric of space..

**Comment by sjaakd** [comment]
@agudian , @gunnarmorling .. I remember why I delegated the `NullValueMappingStrategy` to the forged method. Because in a way, we already supported this. See commit 5825f33. So what do we do?

@agudian .. In retrospect its easier to always add the suffix. Do you mind?

**Comment by agudian** [comment]
> In retrospect its easier to always add the suffix. Do you mind?

No, sounds good!

**Comment by sjaakd** [comment]
> So what do we do?

I can still avoid the situations where we end up in a warning and leave the test case intact. The warnings only applied to the getter only / update methods (so where we use the .addAll at the call site). But what if we delegate the NVMS to the forged method only in the setter scenario? That would fix this. I'll add another commit.

**Comment by sjaakd** [comment]
> So what do we do?

We could also leave: `org.mapstruct.ap.test.collection.forged.CollectionMappingTest.java` as is now. As @gunnarmorling pointed out, the NVMS is intended to work on the bean mapping method result, not on individual properties. So you could also claim that the failing test cases were made with the wrong assumption.

Anyway... WDYT?

**Comment by gunnarmorling** [comment]
>  So you could also claim that the failing test cases were made with the wrong assumption.

What failing tests do you mean specifically?

**Comment by sjaakd** [comment]
> What failing tests do you mean specifically?

checkout commit 5825f33, modication in `org.mapstruct.ap.test.collection.forged.CollectionMappingTest.java`. Expected null (because of NVMS) is now changed to empty.

**Comment by sjaakd** [comment]
@gunnarmorling as discussed yesterday. 

1. Delegation of configuration parameters such as `NullValueMappingStrategy` runs from: config -> type (mapper) -> method (bean-mapping, iterable-mapping, map-mapping).

2. Hence a generated iterable-mapping, map-mapping should take its configuration parameters from either config or type (mapper).

This would lead to including possible NPE's on the call site ( generated property mapping ) before applying the addAll in case of getter only accessor and update methods. There are 2 ways to solve this:

1. Generate a different forged method for these cases (advantage, one check less, cleaner code, better understandable templates).

2. Include a null check at call site.

We decided in the hangout to go for option 1. I stuck to that solution.

---
If there are no further objections I'll continue and merge later today.

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 14]
```diff
@@ -28,6 +28,7 @@
 public enum Message {
 
     // CHECKSTYLE:OFF
+// CHECKSTYLE:OFF
```
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Seems not needed?

Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: The `CHECKSTYLE:OFF` is needed for keeping the messages on one line (violates the line length). However, I did not touch this file functionally anymore..

**Comment by gunnarmorling** [comment]
@sjaakd I've played a bit with the example provided by the issue reporter (`DomainDtoMapper`). If I change the type of `Domain#strings` into `Set<Long>`, a forged iterable mapping method is generated (as expected). This one returns an empty set though if the input is `null`. Shouldn't it return `null` though as the default NVMS is `RETURN_NULL`?

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 14 above)
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Yes, but why have it twice? It's already there a line before :)

Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Hmm. That's a good point :smile:

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 15]
```diff
@@ -701,6 +692,24 @@ private String getName(Type type) {
...
+
+        private NullValueMappingStrategyPrism determineForgedNullValueMappingStrategy() {
+
+            NullValueMappingStrategyPrism nvms;
+
+            if ( method.isUpdateMethod() || TargetWriteAccessorType.GETTER == targetWriteAccessorType ) {
+                nvms = NullValueMappingStrategyPrism.RETURN_DEFAULT;
+            }
+            else {
+                if ( method.getMapperConfiguration().isMapToDefaultOnMapperAndMappingConfigLevel() ) {
```
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: Couldn't you simply do
    nvms = method.getMapperConfiguration().getNullValueCheckStrategy() == NullValueMappingStrategyPrism.RETURN_DEFAULT;

That method `isMapToDefaultOnMapperAndMappingConfigLevel()`, seems a bit dubious to me.

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java [hunk 16]
```diff
@@ -65,17 +63,35 @@ public Assignment getNewCollectionOrMapAssignment() {
     }
 
     @Override
+    public List<Type> getThrownTypes() {
+        List<Type> parentThrownTypes = super.getThrownTypes();
+        List<Type> result = new ArrayList<Type>( parentThrownTypes );
+        for ( Type thrownTypeToExclude : thrownTypesToExclude ) {
+            for ( Type parentThrownType : parentThrownTypes ) {
+                if ( parentThrownType.isAssignableTo( thrownTypeToExclude ) ) {
```
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMaps.java]: Is this one related to the actual issue? And shouldn't this logic be applied in a more central location than only this particular assignment?

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java (hunk 15 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: It was already there.. The `NullValueCheckStrategy` is a different thing than `NullValueMappingStrategy`..

**Comment by sjaakd** [comment]
> @sjaakd I've played a bit with the example provided by the issue reporter (DomainDtoMapper). If I change the type of Domain#strings into Set<Long>, a forged iterable mapping method is generated (as expected). This one returns an empty set though if the input is null. Shouldn't it return null though as the default NVMS is RETURN_NULL?

I'll have a look later. Need to go now..

**Comment by sjaakd** [comment]
> @sjaakd I've played a bit with the example provided by the issue reporter (DomainDtoMapper). If I change the type of Domain#strings into Set, a forged iterable mapping method is generated (as expected).

But it is a `Set` now.. So what do you mean by 'change the type of Domain#strings into Set'?

```java
public class Domain {

    private Set<String> strings = new HashSet<String>();

    public Set<String> getStrings() {
        return strings;
    }

    public void setStrings(Set<String> strings) {
        this.strings = strings;
    }
}

-----

> This one returns an empty set though if the input is null. Shouldn't it return null though as the default NVMS is RETURN_NULL?

So, IIUC you mean this:

```java
public class DomainDtoMapperImpl implements DomainDtoMapper {

    @Override
    public Domain update(Dto source, Domain target) {
        if ( source == null ) {
            return null;
        }

        if ( target.getStrings() != null ) {
            target.getStrings().clear(); // perhaps move this inside check
            List<String> strings = source.getStrings();
            if ( strings != null ) {
                target.getStrings().addAll( strings );
            }
           // start change
           else {
               target.setString( null );
           }
           // end change
        }
        else {
            List<String> strings = source.getStrings();
            if ( strings != null ) {
                target.setStrings( new HashSet<String>( strings ) );
            }
        }

        return target;
    }
}
```

**Comment by gunnarmorling** [comment]
Ah it was escaped. I meant a Set of another type, requiring a Konversion.

**Comment by sjaakd** [comment]
> Ah it was escaped. I meant a Set of another type, requiring a Konversion.

I see.. I'll check this out along with the above.

**Comment by sjaakd** [comment]
Ok. I checkout out your comment. 

The generated code has the same problem as of the (non conversion) original case. Right? It both returns NVMS DEFAULT. 

Perhaps we should indeed always include a null check. But it should be on the result of the forged method. For the forged method I'm still in favor of generating the proper method based on the type NVMS.

But perhaps not based anymore on the fact its an update method / getter-only property / setter-and-getter-property. This requires some more work and adding some test cases. 

--------
Apart from this, I think that a method both being a create and update method at the same time (such as in this example) complicates stuff a lot without bringing user a lot of value. Perhaps we should remove this possibility. WDYT?

**Comment by sjaakd** [comment]
@gunnarmorling do we want to set the target iterable property explicitly no null for update methods?

```java
        if ( target.getStrings() != null ) {
            List<String> strings = source.getStrings();
            if ( strings != null ) {
                target.getStrings().clear(); 
                target.getStrings().addAll( strings );
            }
           else {
               target.setString( null );
           }
        }
```

**Comment by gunnarmorling** [comment]
Yes, I think for an update method that'd make sense. For a normal mapping
method I'd leave the target property alone if the source property is null.

**Comment by gunnarmorling** [comment]
Hey @sjaakd, how is it going with this one? Do you still think we can aim for the CR 2 this week?

**Comment by sjaakd** [comment]
> Yes, I think for an update method that'd make sense. For a normal mapping
> method I'd leave the target property alone if the source property is null.

I leave this one for now.. Need some more refactoring to enable this. Separate PR after release?


> Hey @sjaakd, how is it going with this one? Do you still think we can aim for the CR 2 this week?

Troubling.. Feels like fitting a too large carpet in a too small room. If you push down one corner, the other one pops up. Simply too many combinations.. The culprit is the test case assert commented out in the last commit above. It seems to conflict with #289 test case. The main question: when to clear a target in case of an update method and getter only?

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/Assignment.java [hunk 17]
```diff
@@ -106,4 +107,10 @@
 
 
     boolean isUpdateMethod();
+
+    /**
+     *
+     * @return return whether to return a null check (true) or not (false).
+     */
+    boolean isIncludeNullCheck();
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/Assignment.java]: is not used, currently.. Can be removed again.

**Comment by sjaakd** [comment]
@gunnarmorling ..see my comments above.. I guess its ready for review

**Comment by sjaakd** [comment]
Wrt to my question about the test case adaptation in 562f572, I did some additional investigation.

Original generated code:

```java
        if ( source.getMapValues() != null ) {
            if ( preExistTarget.getMapValues() != null ) {
                preExistTarget.getMapValues().clear();
                Map<String, String> targetMap = source.getMapValues();
                if ( targetMap != null ) {
                    preExistTarget.getMapValues().putAll( targetMap );
                }
            }
        }
```

New code:
```java
        if ( preExistTarget.getMapValues() != null ) {
            preExistTarget.getMapValues().clear();
            Map<String, String> map = source.getMapValues();
            if ( map != null ) {
                preExistTarget.getMapValues().putAll( map );
            }
        }
```

The change is due to `PropertyMapping`, line 492 and following:
```java
            // For collections and maps include a null check, when the assignment type is DIRECT.
            // for mapping methods (builtin / custom), the mapping method is responsible for the
            // null check. Typeconversions do not apply to collections and maps.
            if ( result.getType() == DIRECT ) {
                result = new NullCheckWrapper( result,  getSourcePresenceCheckerRef() );  // << this statement
            }
            else if ( result.getType() == MAPPED && result.isUpdateMethod() ) {
                result = new UpdateNullCheckWrapper( result, getSourcePresenceCheckerRef() );
            }
```

I hava no idea why that code is there. Seems that we do not want to clear the target in case of a direct mapping being null. Do guys know the original reason? Or is this plain wrong? Shall I put this check back?

Note: in the future we might want to reconsider this (when we want to support merging of maps and iterables).

**Comment by sjaakd** [comment]
test this please

**Comment by agudian** [comment]
I vaguely remember that we had feature requests for such things: If you only have a partial structure coming in and want to update an existing object. But I can't remember if we really made its this way.

Because now it seems to be the better alternative would be to handle that on the user side and to either create more specific, slim, bean types for partial updates, or to define a second update method with those properties explicitly excluded.

So let's leave it out and add it to the migration notes?

**Comment by sjaakd** [comment]
> So let's leave it out and add it to the migration notes?

Agreed. I think this stuff has become to tricky. 

@gunnarmorling : I removed the stuff (selecting a proper forged method) again.  So if you agree I want to go ahead and merge.. I'll do some final checking myself this evening.

**Comment by sjaakd** [comment]
@gunnarmorling , @agudian I want to merge this one and finish this.. Any remaining remarks?

**Comment by agudian** [comment]
Nope, go ahead! 👍

**Comment by sjaakd** [comment]
Thanks!

**Comment by gunnarmorling** [comment]
I can't remember the reasoning behind this :/ These things have become quite complex unfortunately. I hope we can find ways to simplify them a bit after 1.1 Final. Thanks for driving this home, @sjaakd!
